### PR TITLE
gh-90928: Improve static initialization of keywords tuple in AC

### PR DESCRIPTION
**Pull request by erlend-aasland**
Calculate `num_keywords` during template generation instead of leaving it to the preprocessor. This makes for much smaller and easier to read (FWIW) AC output.

See discussion in https://github.com/python/cpython/pull/95860#discussion_r943188783

* Issue: gh-90928

### CODE REVIEW

File: Tools/clinic/clinic.py
Comment (ericsnowcurrently) [inline Tools/clinic/clinic.py]: Move this down below the docstring.

File: Tools/clinic/clinic.py
Comment (ericsnowcurrently) [inline Tools/clinic/clinic.py]: We could simplify a little.  Something like this:

```suggestion
    if num_keywords == 0:
        declarations = """
            #if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)
            #  define KWTUPLE (PyObject *)&_Py_SINGLETON(tuple_empty)
            #else
            #  define KWTUPLE NULL
            #endif
            """"
    else:
        declarations = """
            #if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)

            static struct {{
                PyGC_Head _this_is_not_used;
                PyObject_VAR_HEAD
                PyObject *ob_item[%d];
            }} _kwtuple = {{
                .ob_base = PyVarObject_HEAD_INIT(&PyTuple_Type, %d)
                .ob_item = {{ {keywords_py} }},
            }};
            #define KWTUPLE (&_kwtuple.ob_base.ob_base)

            #else  // !Py_BUILD_CORE
            #  define KWTUPLE NULL
            #endif  // !Py_BUILD_CORE
            """ % (num_keywords, num_keywords)

    declarations += """

        static const char * const _keywords[] = {{{keywords_c} NULL}};
        static _PyArg_Parser _parser = {{
            .keywords = _keywords,
            %s
            .kwtuple = KWTUPLE,
        }};
        #undef KWTUPLE
        """ % (format_ or fname)
```

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Thanks for working on this!  Mostly LGTM.

Also, can you drop "num_keywords" from `template_dict` now?

**Comment by erlend-aasland** [comment]
> Also, can you drop "num_keywords" from template_dict now?

It is already dropped :)

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Oops, of course.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Yes, that makes sense.

**Comment by erlend-aasland** [comment]
I have made the requested changes; please review again.

File: Tools/clinic/clinic.py
Comment (ericsnowcurrently) [inline Tools/clinic/clinic.py]: FWIW, you could do away with `keywords()`:

```suggestion
    num_keywords = len([p for p in f.parameters.values()
                        if not p.is_positional_only() and not p.is_vararg()])
```

File: Tools/clinic/clinic.py
Comment (ericsnowcurrently) [inline Tools/clinic/clinic.py]: It would remain helpful to keep `NUM_KEYWORDS` (for readability):

```suggestion
        declarations = """
            #if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)

            #define NUM_KEYWORDS %d
            static struct {{
                PyGC_Head _this_is_not_used;
                PyObject_VAR_HEAD
                PyObject *ob_item[%d];
            }} _kwtuple = {{
                .ob_base = PyVarObject_HEAD_INIT(&PyTuple_Type, NUM_KEYWORDS)
                .ob_item = {{ {keywords_py} }},
            }};
            #undef NUM_KEYWORDS
            #define KWTUPLE (&_kwtuple.ob_base.ob_base)

            #else  // !Py_BUILD_CORE
            #  define KWTUPLE NULL
            #endif  // !Py_BUILD_CORE
        """ % (num_keywords)
```

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

I'll leave it up to you about the two comments I left.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Yeah, the initial draft I had used something like that. I added the nested function for readability. I have no strong preference here; I'm fine with it as it is, and I'm also fine with inlining it.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Yes, this is something I considered. I agree that it improves readability. Thanks.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: We should fix this as well before merging. Give me two seconds...

**Comment by erlend-aasland** [comment]
Thanks for reviewing; highly appreciated. I'll make a quick adjustment before landing.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Fixed in 8915c0550f8b0776545267f7bc9444b87e419786

**Comment by erlend-aasland** [comment]
I'm done with the last round of adjustments. I ended up adding both of your last suggestions. Thanks again.

**Comment by erlend-aasland** [comment]
FYI, I'll wait with merging until Kumar has finished his review.

**Review by kumaraditya303 — APPROVED** [review]
LGTM, thanks!

**Comment by erlend-aasland** [comment]
Thanks for the reviews!

**Pull request merged**
